src/server/channel_store.h

```diff
@@ -23,45 +24,53 @@ class ChannelSlice {
     std::string pattern;
 
     Subscriber(ConnectionContext* cntx, uint32_t tid);
+    Subscriber(uint32_t tid);
     // Subscriber() : borrow_token(0) {}
 
     Subscriber(Subscriber&&) noexcept = default;
     Subscriber& operator=(Subscriber&&) noexcept = default;
 
     Subscriber(const Subscriber&) = delete;
     void operator=(const Subscriber&) = delete;
-  };
 
-  std::vector<Subscriber> FetchSubscribers(std::string_view channel);
+    // Sort by thread-id. Subscriber without owner comes first.
+    static bool ByThread(const Subscriber& lhs, const Subscriber& rhs) {
+      if (lhs.thread_id == rhs.thread_id)
+        return (lhs.conn_cntx != nullptr) < (rhs.conn_cntx != nullptr);
+      return lhs.thread_id < rhs.thread_id;
+    }
+  };
 
   void AddSubscription(std::string_view channel, ConnectionContext* me, uint32_t thread_id);
   void RemoveSubscription(std::string_view channel, ConnectionContext* me);
 
   void AddGlobPattern(std::string_view pattern, ConnectionContext* me, uint32_t thread_id);
   void RemoveGlobPattern(std::string_view pattern, ConnectionContext* me);
 
+  std::vector<Subscriber> FetchSubscribers(std::string_view channel);
+
   std::vector<std::string> ListChannels(const std::string_view pattern) const;
   size_t PatternCount() const;
 
  private:
-  struct SubscriberInternal {
-    uint32_t thread_id;  // proactor thread id.
+  using SubscribeMap = absl::flat_hash_map<ConnectionContext*, unsigned>;
 
-    SubscriberInternal(uint32_t tid) : thread_id(tid) {
-    }
+  struct ChannelMap : absl::flat_hash_map<std::string, SubscribeMap> {
```

Comment:
it can be pretty heavy op when the main map does rehashing. and it runs under lock. Consider using pointers in values

Reply:
Why? Do reduce memory movement in rehashing?

Reply:
Either way, did it